Periodically sweep the IPCache map on kernels that don't support delete operation by recreating the map with the desired entries

joestringer opened the pull request:
Add detection of the case where Linux kernels 4.11 to 4.14 inclusive don't support delete operations on IPCache (LPM) maps.

When this case is detected, change the IPCache garbage collection to instead:
* Grab a readlock on the IPCache
* Create a brand new map on the filesystem
* Populate the new map based on the current IPCache state
* Move the old map out of the way
* Move the new map into the standard IPCache location
* Reload BPF programs (Should not require recompilation, but it performs recompilation for now)
* When successful, delete the old map from the filesystem.
  * If not successful, attempt to roll back

This should be safe, because when a BPF program is loaded, it reads the filesystem at that point and takes a reference on the map that is pinned to the standard location. It continues to use this reference to the map that was loaded and pinned to that location for the remainder of the lifetime of the BPF program.

When we prepare the new map, even when we move it into the same location as the previous map, the existing BPF programs continue to use the old map based on the reference from load time. By placing it into the location that was previously used, the path does not need to change in the existing BPF ELF binary so no recompilation is necessary. When the BPF program is reloaded from the same ELF, it looks into the filesystem and gets a reference to the new map, and we discard the reference to the old map when replacing the previous program's attachment. Finally, we reload the IPCache map from the Cilium userspace side (at the same path again) to remove the final reference to the old IPCache map.

Fixes: #4896

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5303)
<!-- Reviewable:end -->

rlenglet commented on pkg/datapath/ipcache/listener.go:
```diff
@@ -111,48 +156,8 @@ func (l *IPCacheListenerBPF) OnIPIdentityCacheGC() {
-				return nil
-			},
+			DoFunc:      l.garbageCollect,
 			RunInterval: time.Duration(5) * time.Minute,
```
time.Duration(5) -> 5

rlenglet reviewed (commented):
The ipcache commits LGTM.

joestringer committed:
maps/ipcache: Add map Reopen() function

This will be used by an upcoming commit.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
ipcache: Add function to dump contents of IPCache

Add a new DumpToListenerLocked() function which calls the listener's
OnIPIdentityChange method for every element in the IPCache. This will be
used by an upcoming commit.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
maps/ipcache: Add mechanism to detect delete support

ipcache.SupportsDelete() blocks until a delete occurs, then returns
whether the deletion was successful or not. This can be used to
determine whether periodic deletion of the IPCache map on the filesystem
is necessary or not.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
datapath/ipcache: Indent GC code

This reduces the LoC diff for the following patch. No functional change.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
datapath/ipcache: GC maps when delete is unsupported

On Linux versions 4.11 to 4.14 inclusive, the LPM map type doesn't
support delete. To manage this, in the current pkg/maps/ipcache we just
overwrite the "deleted" entries with zeroed out values. This is also
handled in the datapath.

However, over time these entries will take up space and potentially
prevent newer IPCache entries from taking over.

To avoid this resource leakage, detect support for the LPM delete
operation, and if it is unsupported, delete the old map and populate a
brand new map in its place based on the latest set of entries in the
IPCache.

Fixes: #4896

Signed-off-by: Joe Stringer <joe@covalent.io>

tgraf commented on daemon/daemon.go:
```diff
@@ -1462,6 +1462,23 @@ func mapValidateWalker(path string) error {
+//
+// The returned WaitGroup allows the caller to wait until all endpoints have
+// finished regenerating.
+func (d *Daemon) TriggerReload(context string) (*sync.WaitGroup, error) {
```
Can you make the naming here more explicit? This really only compiles the base programs and *reloads* the endpoint programs without recompiling?

joestringer replied:
That's the intention, but it's not optimized to skip the compile yet (I remember @ianvernon briefly discussing adding that "skip compile" piece as one of his PRs, but I'll sync with him on the status).

I can rename this to `TriggerReloadWithoutRecompile()`, perhaps?

joestringer replied:
Do you have a recommendation for a more explicit name? Right now I'm leaning more towards something like `ReloadDatapath()` or `ReloadBPF()` because the point of this function is that it requires Reload, but not necessarily compile/policy recalculation/etc. Whether or not the underlying code is efficient enough to avoid compile/policy recalculation/etc is a matter for the implementation, not for the callers.

(I'll update the comment, too)

joestringer replied:
I believe that we only need this patch to achieve the above optimization of skipping compile from this path, on top of the existing PR here:

https://github.com/joestringer/cilium/commit/f07ce4e2bbc24f8c37a795e2401d74a0b5af8d6c

But I'd rather split it into its own PR for closer review.

tgraf replied:
I would appreciate the `WithoutRecompile` personally but I'm also fine with crystal clear function doc

joestringer replied:
On one hand, I could see how the `WithoutCompile` part could get out of sync with the underlying implementation.

On the other hand, I can also see how for future (potential) users of the function, it would be important to have that part in the name, to determine if it's a safe version to use for triggering regeneration. 

I can add that in after I see another green build (so nothing broke during rebase).

joestringer committed:
endpoint: Refactor regeneration context

Create a new structure to hold the context for regeneration. Initially,
this will hold just the reason which can be used for logging. Upcoming
commits will also allow the caller to set a bool inside the structure to
force reloading of the BPF datapath programs.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
daemon: Add support for datapath reload

Create a new function for triggering unconditional reload of all
programs without needing to recalculate policy or regenerate the
datapath BPF programs. This will be used by an upcoming commit that
replaces the IPCache map in-place, which requires reload (but not
recompilation) to switch to the new map.

Signed-off-by: Joe Stringer <joe@covalent.io>

ianvernon commented on daemon/policy.go:
```diff
@@ -51,7 +51,8 @@ func (d *Daemon) TriggerPolicyUpdates(force bool, reason string) *sync.WaitGroup
 		log.Debugf("Full policy recalculation triggered")
 	}
-	return endpointmanager.RegenerateAllEndpoints(d, reason)
+	regenContext := endpoint.NewRegenerationContext(reason)
```
Hooray!!! This made my day - in the future, we can eventually take into account what the contents of these regenerations are and be more selective in how we regenerate :D

ianvernon commented on pkg/endpoint/policy.go:
```diff
@@ -67,6 +67,23 @@ var (
 	}
 )
 
+// RegenerationContext provides context to regenerate() calls to determine
```
Nit: can you put this in its own file?

joestringer replied:
Can be done, I wasn't sure what the threshold should be for putting things in their own file since there isn't really any code associated directly with the `RegenerationContext`.

borkmann reviewed (approved):
lgtm

joestringer commented on pkg/maps/ipcache/ipcache.go:
```diff
@@ -159,18 +169,33 @@ func NewMap(name string) *Map {
+		if overwrite {
+			return false, m.Update(k, &RemoteEndpointInfo{})
+		}
+		return false, nil
```
technically this should be `false, err` but it doesn't matter for existing users.
